[quant][graphmode] Preserve numerics in debug option for clamp ops

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39219 [quant][graphmode] Preserve numerics in debug option for clamp ops**

Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

jerryzh168 committed:
[quant][graphmode] Preserve numerics in debug option for clamp ops

Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -624,13 +701,28 @@ ModuleMethodVector InsertQuantDeQuantHelper::getInvokedMethods(
   Node* n = original_output->node();
   Graph* graph = n->owningGraph();
+  if (is_scalar) {
+    // convert Scalar to Tensor
```
Might be useful to make a scalar to tensor and tensor to scalar conversion function for this snippet and reuse it across ops.

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -339,6 +351,46 @@ void RemoveRedundantQuantizationOps(std::shared_ptr<Graph>& graph) {
   rewriter.runOnGraph(graph, filter);
 }
 
+void ReplicateClampScalarArgs(std::shared_ptr<Graph>& graph) {
```
What are you replicating here? Why do you need to replicate the min/max args?

jerryzh168 replied:
this is to ensure these args can be matched in patterns

raghuramank100 reviewed (commented):
Requesting clarity on a few questions

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1371,7 +1352,8 @@ def forward(self, x):
             op = "quantized::add_scalar" if quantized else "aten::add"
-            m = self._test_op_impl(m, data, op)
+            # TODO: fix debug=True numerics
+            m = self._test_op_impl(m, data, op, check=False)
```
All the op impls call the same model m, which gets called with quantize_script twice. Can we refactor the tests so that we call quantize_script on the traced and scripted model with debug=True/False and then just check for the quantized op being present/absent?

jerryzh168 replied:
adding tracing support will be a separate effort as well, maybe we'll need more help to cover tracing on all ops

raghuramank100 replied:
In that case, lets add support for scripting first and refactor the test. Let make a list for what ops need tracing support and follow up in later PRs and track the work.

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1113,20 +1087,27 @@ class TestQuantizeScriptPTSQOps(QuantizationTestCase):
     for individual ops end to end.
     """
-    def _test_op_impl(self, module, data, quantized_op, debug=False):
+    def _test_op_impl(self, module, data, quantized_op, check=True):
```
Suggest refactoring this into two parts:
1. Script/Trace the model and create the outputs dict. 
2. Check for specific ops in the output dict

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1371,7 +1352,8 @@ def forward(self, x):
                              (NonQuantizedInplaceAddScalar(), False)]:
             op = "quantized::add_scalar" if quantized else "aten::add"
-            m = self._test_op_impl(m, data, op)
+            # TODO: fix debug=True numerics
```
For compactness, can we collapse the 4 models per op into a single model? Can we have a single forward doing both non-inplace and inplace operations? Since you duplicate this structure for other ops, this can make the code easy to read.

jerryzh168 replied:
I can try it in a separate PR, I'm not sure if that is scriptable

raghuramank100 commented on torch/csrc/jit/passes/quantization/finalize.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <torch/csrc/jit/passes/quantization/finalize.h>
+#include <torch/csrc/jit/jit_log.h>
```
Is this change part of this PR?

jerryzh168 replied:
this is not needed anymore

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -142,6 +145,31 @@ Node* insertQParam(
   return qparam;
 }
 
+Node* insertScalarToTensor(Graph* graph, Value* scalar_value) {
```
These seem general purpose, should they be in one of the helper files?

jerryzh168 replied:
this is only used in insert_quant_dequant

raghuramank100 commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -34,27 +35,78 @@ std::string getExtraArgList(std::vector<std::string> extra_args) {
   return aten_op_pattern;
 }
 
+std::string getQuantize(const std::string& value) {
```
can you add comments to explain what this function does?

raghuramank100 commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -34,27 +35,78 @@ std::string getExtraArgList(std::vector<std::string> extra_args) {
+std::string getQuantize(const std::string& value) {
+  std::string quantize_pattern = R"(
+          )" +
+      value + "_float_scalar_type : int = prim::Constant[value=6]()";
```
Why 6 here?

jerryzh168 replied:
6 is `torch.float`

raghuramank100 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -704,7 +817,7 @@ void propagateDequantize(Value* output, const std::vector<Value*> inputs) {
 }
 
-void propagateQuantizationOps(Block* block) {
+void InsertQuantDeQuantHelper::propagateQuantizationOps(Block* block) {
```
Can you add a comment stating for which ops we do propagation with this function?

jerryzh168 replied:
added in declaration

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Preserve numerics in debug option for clamp ops"


Summary:
We didn't model clamp ops correctly right now, this PR fixes that.

Reason is quantized clamp op quantizes the scalar arguments in the op implementation: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L614-L617

So we'll need to model this explicitly in the IR.
When we see a `aten::dequantize - aten::clamp(%x, %min, %max)`
we first make a scalar tensor with `aten::scalar_tensor(%scalar, ...)`, then we quantize the tensor with the same quantization parameters from the input tensor of the `aten::clamp`, dequantize the tensor, then convert the dequantized tensor to scalar using `aten::item`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21831350](https://our.internmc.facebook.com/intern/diff/D21831350)

[ghstack-poisoned]
